### PR TITLE
Allowing nil to represent an empty slice for generated fields.

### PR DESCRIPTION
**Pull request by shirchen**
Relaxing previous restriction where slices generated for required fields of a struct must be non-nil  if empty. Apache Thrift does not have this restriction so making migration easier.

### CODE REVIEW

File: gen/list.go [hunk 1]
```diff
@@ -55,6 +55,9 @@ func (l *listGenerator) ValueList(g Generator, spec *compile.ListSpec) (string,
 			<$f := newVar "f">
 			<$w := newVar "w">
 			func (<$v> <.Name>) ForEach(<$f> func(<$wire>.Value) error) error {
+				if <$v> == nil {
+					return nil
```
Comment (abhinav) [inline gen/list.go]: not necessary anymore

File: gen/field.go [hunk 2]
```diff
@@ -255,9 +255,11 @@ func (f fieldGroupGenerator) ToWire(g Generator) error {
 				<- $f := printf "%s.%s" $v $fname ->
 				<- if .Required ->
 					<- if not (isPrimitiveType .Type) ->
-						if <$f> == nil {
-							return <$wVal>, <import "errors">.New("field <$fname> of <$structName> is required")
-						}
+						<- if not (isListType .Type) ->
```
Comment (abhinav) [inline gen/field.go]: can use `and (not (..)) (not (..))`/`not (or (..) (..))`

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,7 +5,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
-- No changes yet.
+- Allowing nil to represent an empty slice for generated fields. Relaxing previous
```
Comment (abhinav) [inline CHANGELOG.md]:
```
### Changed
- `nil` slices are now treated as empty lists for fields of `list` type. This
  relaxes the previous requirement of accepting only non-`nil` slices for
  required `list` fields.
```

File: gen/container_test.go [hunk 4]
```diff
@@ -1306,14 +1312,16 @@ func TestContainerValidate(t *testing.T) {
 	}
 
 	for _, tt := range tests {
-		value, err := tt.value.ToWire()
-		if err == nil {
-			err = wire.EvaluateValue(value) // lazy error
-		}
+		t.Run(tt.desc, func(t *testing.T) {
```
Comment (abhinav) [inline gen/container_test.go]: thanks for doing this

File: gen/internal/tests/thrift/containers.thrift [hunk 5]
```diff
@@ -42,6 +42,9 @@ struct MapOfBinaryAndString {
     1: optional map<binary, string> binaryToString;
     2: optional map<string, binary> stringToBinary;
 }
+struct ListOfPrimitives {
```
Comment (abhinav) [inline gen/internal/tests/thrift/containers.thrift]: nit: empty line between declarations

```suggestion

struct ListOfPrimitives {
```

File: gen/container_test.go [hunk 6]
```diff
@@ -1364,6 +1372,29 @@ func TestEmptyContainersRoundTrip(t *testing.T) {
 		assert.Equal(t, give, got)
 	})
 
+	t.Run("required with special treatment of nil slice", func(t *testing.T) {
```
Comment (abhinav) [inline gen/container_test.go]: can we also add an explicit test to verify that we continue to differentiate between nil and empty for optional list fields?

File: gen/roundtrip_test.go [hunk 7]
```diff
@@ -36,41 +36,51 @@ import (
...
+	gotX := reflect.New(xType).Interface().(thriftType)
+	if assert.NoError(t, gotX.FromWire(v), "FromWire: %v", message) {
+		return assert.Equal(t, x, gotX, "FromWire: %v", message)
+	}
+
+	return false
+}
+
+// assertSerialization checks if x.ToWire() results in the provided wire.Value.
+func assertSerialization(t *testing.T, x thriftType, v wire.Value, message string) (bool, wire.Value) {
```
Comment (abhinav) [inline gen/roundtrip_test.go]: Let's do ToWire in the caller.

```
assertBinaryRoundTrip(t *testing.T, v wire.Value, message string) (wire.Value, bool)
```

**Comment by abhinav** [comment]
For the validity test here, I don't think we should round trip. What we want to validate is:

1. ToWire
    1. Struct with required `list` field as input
        1. `list` is nil: should encode into a `wire.Value` that holds an
            empty list. (Stop at `wire.ValuesAreEqual`.)
        2. `list` is empty: should also encode into a `wire.Value` that holds
            an empty list.
    2. Struct with optional `list` field as input
        1. `list` is nil: should encode into a `wire.Value` that DOES NOT hold
            the list field.
        2. `list` is empty: should also encode into a `wire.Value` that holds
            an empty list.
2. FromWire
    1. Decode a `wire.Value` into a struct with a required `list` field
        1. `wire.Value` has an empty list field: decodes into empty
        2. `wire.Value` is missing the list field: fails
    2. Decode a `wire.Value` into a struct with an optional `list` field
        1. `wire.Value` has an empty list field: decodes into empty
        2. `wire.Value` is missing the list field: decodes to nil

File: gen/list_test.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+		desc       string
+		p          *tc.ListOfRequiredPrimitives
+		offTheWire wire.Value
+		expect     []string
+	}{
+		{
+			"required nil list: wire.Value with empty list",
+			&tc.ListOfRequiredPrimitives{
+				ListOfStrings: nil,
+			},
```
Comment (abhinav) [inline gen/list_test.go]: Would you mind using named fields here? Positional works fine for 2 fields but after that we should just do named.

(I know a lot of existing tests in ThriftRW don't do that but they're really old.)

File: gen/list_test.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+//      wire.Value is missing the list field: decodes to nil.
+//
+// We also test that full round trip works as expected with protocol en/decoding.
+
+// TestListRequired tests that empty slice and nil allocate an empty list after deserialization via
+// FromWire.
+func TestListRequiredToWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		p          *tc.ListOfRequiredPrimitives
```
Comment (abhinav) [inline gen/list_test.go]: nit:

```suggestion
		give          *tc.ListOfRequiredPrimitives
```

File: gen/list_test.go [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+		offTheWire wire.Value
```
Comment (abhinav) [inline gen/list_test.go]: nit:

```suggestion
		want wire.Value
```

File: gen/list_test.go [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+				{
+					ID:    1,
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			[]string{},
+		},
+	}
+	for _, tt := range tests {
+		// required
```
Comment (abhinav) [inline gen/list_test.go]: subtests here:

    t.Run(tt.desc, func(t *testing.T) { .. })

File: gen/list_test.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+		w, err := tt.p.ToWire()
+		assert.NoError(t, err, "failed to serialize: %encodedValue", tt.p)
```
Comment (abhinav) [inline gen/list_test.go]: If there was an error here, we can't proceed with the rest of the test for
this case because the following code assumes that it has a valid wire.Value.

```suggestion
		require.NoError(t, err, "failed to serialize: %encodedValue", tt.p)
```

File: gen/list_test.go [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+		assert.True(t, wire.ValuesAreEqual(tt.offTheWire, w))
+
+		// Round trip them all.
+		freshestV, b := assertBinaryRoundTrip(t, w, tt.desc)
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
		got, ok := assertBinaryRoundTrip(t, w, tt.desc)
```

File: gen/list_test.go [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+		assert.True(t, b, "failed round trip")
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
		require.True(t, b, "failed round trip")
```

File: gen/list_test.go [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+		assert.True(t, tt.p.Equals(tt.p))
+		// Allocate a new instance to serialize from Thrift representation.
+		x := new(tc.ListOfRequiredPrimitives)
+		if assert.NoError(t, x.FromWire(freshestV), tt.desc) {
```
Comment (abhinav) [inline gen/list_test.go]: no need to `tt.desc` here because if this assertion fails, it'll be connected to the subtest's name.

File: gen/list_test.go [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+
+func TestListRequiredFromWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
+		expected   *tc.ListOfRequiredPrimitives
+	}{
+		{
+			"empty list field decodes into an empty slice",
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
```
Comment (abhinav) [inline gen/list_test.go]: named fields please

File: gen/list_test.go [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+		if assert.NoError(t, x.FromWire(freshestV), tt.desc) {
+			assert.Equal(t, tt.expect, x.ListOfStrings)
+		}
+	}
+}
+
+func TestListRequiredFromWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
		give wire.Value
```

File: gen/list_test.go [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+		expected   *tc.ListOfRequiredPrimitives
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
		want   *tc.ListOfRequiredPrimitives
```

File: gen/list_test.go [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			&tc.ListOfRequiredPrimitives{
+				ListOfStrings: []string{},
+			},
+		},
+	}
+	for _, tt := range tests {
+		x := new(tc.ListOfRequiredPrimitives)
```
Comment (abhinav) [inline gen/list_test.go]: t.Run(tt.desc

File: gen/list_test.go [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+		if assert.NoError(t, x.FromWire(tt.offTheWire), tt.desc) {
+			assert.Equal(t, tt.expected, x)
+			_, b := assertBinaryRoundTrip(t, tt.offTheWire, tt.desc)
+			assert.True(t, b, "failed round trip")
+		}
+	}
+}
+
+// Error if required list is missing in the wire representation.
+func TestListRequiredFromWireError(t *testing.T) {
```
Comment (abhinav) [inline gen/list_test.go]: For this and TestListRequiredFromWire, since there's only one case each, we can do,

```
func TestListRequiredFromWire(t *testing.T) {
  t.Run("empty list decodes into an empty slice", func(t *testing.T) {
    give := wire.NewValueList(..)
    want := tc.ListOfPrimitives{...}
  
    var got tc.ListOfRequiredPrimitives
    require.NoError(t, got.FromWire(give), "failed to decode")
    assert.Equal(t, want, got, "decoded value did not match")
    
    _, ok := assertBinaryRoundTrip(..)
    assert.True(t, ok, "failed to round trip")
  })
  
  t.Run("absent list field results in an error", func(t *testing.T) {
    ...
  })
}
```

File: gen/list_test.go [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+}
+
+// TestListOptionalToWire tests optional serialization cases.
+func TestListOptionalToWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
+		p          *tc.ListOfOptionalPrimitives
+	}{
+		{
```
Comment (abhinav) [inline gen/list_test.go]: named fields please

File: gen/list_test.go [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+			assert.Equal(t, tt.wantError, err.Error())
+		}
+	}
+}
+
+// TestListOptionalToWire tests optional serialization cases.
+func TestListOptionalToWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
		give wire.Value
```

File: gen/list_test.go [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+		p          *tc.ListOfOptionalPrimitives
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
		want          *tc.ListOfOptionalPrimitives
```

File: gen/list_test.go [hunk 24]
```diff
@@ -0,0 +1,220 @@
...
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			&tc.ListOfOptionalPrimitives{
+				ListOfStrings: []string{},
+			},
+		},
+	}
+	for _, tt := range tests {
+		assertRoundTrip(t, tt.p, tt.offTheWire, tt.desc)
```
Comment (abhinav) [inline gen/list_test.go]: This doesn't need to be a round-trip test. It can be,

    t.Run(tt.desc, func(t *testing.T) {
        var got tc.ListOfOptionalPrimitives
        require.NoError(t, got.FromWire(tt.give), "failed to decode")
        assert.Equal(t, tt.want, got, "values did not match")
    })

The contract we're testing here stops at the wire.Value. It doesn't hurt to
round-trip it, but it's not necessary.

File: gen/list_test.go [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+	}
+}
+
+func TestListOptionalFromWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
		give wire.Value
```

File: gen/list_test.go (hunk 24 above)
Comment (shirchen) [reply gen/list_test.go]: yup, i will leave a comment how full round-trip does more than what we need.

File: gen/container_test.go [hunk 26]
```diff
@@ -1306,14 +1312,16 @@ func TestContainerValidate(t *testing.T) {
 	}
 
 	for _, tt := range tests {
-		value, err := tt.value.ToWire()
-		if err == nil {
-			err = wire.EvaluateValue(value) // lazy error
-		}
+		t.Run(tt.desc, func(t *testing.T) {
+			value, err := tt.value.ToWire()
+			if err == nil && value.String() != "" {
```
Comment (abhinav) [inline gen/container_test.go]: `value.String` check unnecessary. The implementation of `wire.Value.String()` either returns a non-empty string or panics.

```suggestion
			if err == nil {
```

File: gen/list_test.go [hunk 27]
```diff
@@ -0,0 +1,202 @@
...
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			wantList: []string{},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			w, err := tt.give.ToWire()
+			require.NoError(t, err, "failed to serialize: %encodedValue", tt.give)
```
Comment (abhinav) [inline gen/list_test.go]: Typo: `%encodedValue` isn't a valid formatting argument.

```suggestion
			require.NoError(t, err, "failed to serialize", tt.give)
```

File: gen/list_test.go [hunk 28]
```diff
@@ -0,0 +1,202 @@
...
+			require.True(t, wire.ValuesAreEqual(tt.want, w))
+			assert.True(t, tt.give.Equals(tt.give))
+			// Round trip them all.
+			got, b := assertBinaryRoundTrip(t, w, tt.desc)
```
Comment (abhinav) [inline gen/list_test.go]: I think my prior comment about this got hidden so repeating.

Please call the boolean here indicating success `ok`. `ok` is a convention for
that, `b` is less meaningful. Readers will have to check the implementation of
`assertBinaryRoundTrip` to see what `b` means.

```suggestion
			got, ok := assertBinaryRoundTrip(t, w, tt.desc)
```

File: gen/list_test.go [hunk 29]
```diff
@@ -0,0 +1,202 @@
...
+			require.True(t, b, "failed round trip")
+			// Allocate a new instance to serialize from Thrift representation.
```
Comment (abhinav) [inline gen/list_test.go]: Typo:

```suggestion
			// Allocate a new instance to deserialize from Thrift representation.
```

File: gen/list_test.go [hunk 30]
```diff
@@ -0,0 +1,202 @@
...
+			x := new(tc.ListOfRequiredPrimitives)
+			require.NoError(t, x.FromWire(got), tt.desc)
```
Comment (abhinav) [inline gen/list_test.go]: `tt.desc` message for `require.NoError` not needed because `tt.desc` is already the name of this subtest.

File: gen/list_test.go [hunk 31]
```diff
@@ -0,0 +1,202 @@
...
+		give := wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+			{
+				ID:    1,
+				Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+			},
+		}})
+		got := new(tc.ListOfRequiredPrimitives)
+		require.NoError(t, got.FromWire(give), "failed to decode")
+		require.Equal(t, want, got)
+		_, b := assertBinaryRoundTrip(t, give, "empty list field decodes into an empty slice")
```
Comment (abhinav) [inline gen/list_test.go]: Can use `[t.Name](https://golang.org/pkg/testing/#T.Name)` to get the name here instead of duplicating it.
Also, as before, `b` => `ok`.

```suggestion
		_, ok := assertBinaryRoundTrip(t, give, t.Name())
```

File: gen/list_test.go [hunk 32]
```diff
@@ -0,0 +1,202 @@
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			w, err := tt.give.ToWire()
+			require.NoError(t, err, "failed to serialize: %encodedValue", tt.give)
+			require.True(t, wire.ValuesAreEqual(tt.want, w))
+			assert.True(t, tt.give.Equals(tt.give))
+			// Round trip them all.
+			got, b := assertBinaryRoundTrip(t, w, tt.desc)
+			require.True(t, b, "failed round trip")
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
			require.True(t, ok, "failed round trip")
```

File: gen/list_test.go [hunk 33]
```diff
@@ -0,0 +1,202 @@
...
+			{
+				ID:    1,
+				Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+			},
+		}})
+		got := new(tc.ListOfRequiredPrimitives)
+		require.NoError(t, got.FromWire(give), "failed to decode")
+		require.Equal(t, want, got)
+		_, b := assertBinaryRoundTrip(t, give, "empty list field decodes into an empty slice")
+		assert.True(t, b, "failed round trip")
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
		assert.True(t, ok, "failed round trip")
```

File: gen/list_test.go [hunk 34]
```diff
@@ -0,0 +1,202 @@
...
+		give := wire.NewValueStruct(wire.Struct{})
+		x := new(tc.ListOfRequiredPrimitives)
+		err := x.FromWire(give)
+		require.Error(t, err, "failed to decode")
+		assert.Equal(t, "field ListOfStrings of ListOfRequiredPrimitives is required", err.Error())
+	})
+}
+
+// TestListOptionalToWire tests optional serialization cases.
+func TestListOptionalToWire(t *testing.T) {
```
Comment (abhinav) [inline gen/list_test.go]: This isn't a ToWire, but a general optional list field test.

```suggestion
func TestRoundtripOptionalListFields(t *testing.T) {
```

File: gen/list_test.go [hunk 35]
```diff
@@ -0,0 +1,202 @@
...
+	t.Run("absent list field results in an error", func(t *testing.T) {
+		give := wire.NewValueStruct(wire.Struct{})
+		x := new(tc.ListOfRequiredPrimitives)
+		err := x.FromWire(give)
+		require.Error(t, err, "failed to decode")
+		assert.Equal(t, "field ListOfStrings of ListOfRequiredPrimitives is required", err.Error())
+	})
+}
+
+// TestListOptionalToWire tests optional serialization cases.
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
// TestRoundtripOptionalListFields tests optional serialization cases.
```

File: gen/list_test.go [hunk 36]
```diff
@@ -0,0 +1,202 @@
...
+			give: wire.NewValueStruct(wire.Struct{}),
+			want: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			x := new(tc.ListOfOptionalPrimitives)
+			require.NoError(t, x.FromWire(tt.give), tt.desc)
+			assert.Equal(t, tt.want, x.ListOfStrings)
+			_, b := assertBinaryRoundTrip(t, tt.give, tt.desc)
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
			_, ok := assertBinaryRoundTrip(t, tt.give, tt.desc)
```

File: gen/list_test.go [hunk 37]
```diff
@@ -0,0 +1,202 @@
...
+			assert.True(t, b, "failed round trip")
```
Comment (abhinav) [inline gen/list_test.go]:
```suggestion
			assert.True(t, ok, "failed round trip")
```

File: gen/roundtrip_test.go [hunk 38]
```diff
@@ -69,8 +56,28 @@ func assertRoundTrip(t *testing.T, x thriftType, v wire.Value, msg string, args
...
+		return assert.Equal(t, x, gotX, "FromWire: %encodedValue", message)
 	}
+
 	return false
 }
+
+// assertBinaryRoundTrip checks that De/Encode returns the same value.
+func assertBinaryRoundTrip(t *testing.T, w wire.Value, message string) (wire.Value, bool) {
+	var buff bytes.Buffer
+	if !assert.NoError(t, protocol.Binary.Encode(w, &buff), "%encodedValue: failed to serialize", message) {
```
Comment (abhinav) [inline gen/roundtrip_test.go]: typo: `%encodedValue` isn't a thing.

```suggestion
	if !assert.NoError(t, protocol.Binary.Encode(w, &buff), "%v: failed to serialize", message) {
```

File: gen/roundtrip_test.go [hunk 39]
```diff
@@ -69,8 +56,28 @@ func assertRoundTrip(t *testing.T, x thriftType, v wire.Value, msg string, args
...
+		return w, false
+	}
+
+	newV, err := protocol.Binary.Decode(bytes.NewReader(buff.Bytes()), w.Type())
+	if !assert.NoError(t, err, "%encodedValue: failed to deserialize", message) {
```
Comment (abhinav) [inline gen/roundtrip_test.go]: typo:

```suggestion
	if !assert.NoError(t, err, "%v: failed to deserialize", message) {
```

File: gen/roundtrip_test.go [hunk 40]
```diff
@@ -36,31 +36,18 @@ import (
...
-		if !assert.NoError(t, protocol.Binary.Encode(w, &buff), "%v: failed to serialize", message) {
-			return false
-		}
-
-		// Flip v to deserialize(serialize(x.ToWire())) to ensure full round
-		// tripping
 
-		newV, err := protocol.Binary.Decode(bytes.NewReader(buff.Bytes()), v.Type())
-		if !assert.NoError(t, err, "%v: failed to deserialize", message) {
+	if w, err := x.ToWire(); assert.NoError(t, err, "failed to serialize: %encodedValue", x) {
```
Comment (abhinav) [inline gen/roundtrip_test.go]:
```suggestion
	if w, err := x.ToWire(); assert.NoError(t, err, "failed to serialize: %v", x) {
```

File: gen/roundtrip_test.go [hunk 41]
```diff
@@ -36,31 +36,18 @@ import (
...
+		if !assert.True(
+			t, wire.ValuesAreEqual(v, w), "%encodedValue: %encodedValue.ToWire() != %encodedValue", message, x, v) {
```
Comment (abhinav) [inline gen/roundtrip_test.go]:
```suggestion
			t, wire.ValuesAreEqual(v, w), "%v: %v.ToWire() != %v", message, x, v) {
```

File: gen/roundtrip_test.go [hunk 42]
```diff
@@ -36,31 +36,18 @@ import (
...
 			return false
 		}
-
-		if !assert.True(
-			t, wire.ValuesAreEqual(newV, v), "%v: deserialize(serialize(%v.ToWire())) != %v", message, x, v) {
+		// Flip encodedValue to deserialize(serialize(x.ToWire())) to ensure full round trip.
```
Comment (abhinav) [inline gen/roundtrip_test.go]:
```suggestion
		// Flip v to deserialize(serialize(x.ToWire())) to ensure full round trip.
```

File: gen/roundtrip_test.go [hunk 43]
```diff
@@ -36,31 +36,18 @@ import (
...
+		freshV, b := assertBinaryRoundTrip(t, w, message)
+		if !assert.True(t, b, "%encodedValue: failed encode/decode round trip for (%encodedValue.ToWire())) != %encodedValue", x, v) {
```
Comment (abhinav) [inline gen/roundtrip_test.go]:
```suggestion
		if !assert.True(t, b, "%v: failed encode/decode round trip for (%v.ToWire())) != %v", message, x, v) {
```

File: gen/roundtrip_test.go [hunk 44]
```diff
@@ -69,8 +56,28 @@ func assertRoundTrip(t *testing.T, x thriftType, v wire.Value, msg string, args
 	}
 
 	gotX := reflect.New(xType).Interface().(thriftType)
-	if assert.NoError(t, gotX.FromWire(v), "FromWire: %v", message) {
-		return assert.Equal(t, x, gotX, "FromWire: %v", message)
+	if assert.NoError(t, gotX.FromWire(v), "FromWire: %encodedValue", message) {
+		return assert.Equal(t, x, gotX, "FromWire: %encodedValue", message)
```
Comment (abhinav) [inline gen/roundtrip_test.go]:
```suggestion
		return assert.Equal(t, x, gotX, "FromWire: %v", message)
```

File: gen/roundtrip_test.go [hunk 45]
```diff
@@ -69,8 +56,28 @@ func assertRoundTrip(t *testing.T, x thriftType, v wire.Value, msg string, args
 	}
 
 	gotX := reflect.New(xType).Interface().(thriftType)
-	if assert.NoError(t, gotX.FromWire(v), "FromWire: %v", message) {
-		return assert.Equal(t, x, gotX, "FromWire: %v", message)
+	if assert.NoError(t, gotX.FromWire(v), "FromWire: %encodedValue", message) {
```
Comment (abhinav) [inline gen/roundtrip_test.go]:
```suggestion
	if assert.NoError(t, gotX.FromWire(v), "FromWire: %v", message) {
```

File: gen/roundtrip_test.go [hunk 46]
```diff
@@ -36,31 +36,18 @@ import (
...
-
-		newV, err := protocol.Binary.Decode(bytes.NewReader(buff.Bytes()), v.Type())
-		if !assert.NoError(t, err, "%v: failed to deserialize", message) {
-			return false
-		}
-
-		if !assert.True(
-			t, wire.ValuesAreEqual(newV, v), "%v: deserialize(serialize(%v.ToWire())) != %v", message, x, v) {
+		// Flip v to deserialize(serialize(x.ToWire())) to ensure full round trip.
+		freshV, b := assertBinaryRoundTrip(t, w, message)
```
Comment (abhinav) [inline gen/roundtrip_test.go]:
```suggestion
		freshV, ok := assertBinaryRoundTrip(t, w, message)
```

File: gen/roundtrip_test.go [hunk 47]
```diff
@@ -36,31 +36,18 @@ import (
...
+		if !assert.True(t, b, "%v: failed encode/decode round trip for (%v.ToWire())) != %v", x, v) {
```
Comment (abhinav) [inline gen/roundtrip_test.go]:
```suggestion
		if !assert.True(t, ok, "%v: failed encode/decode round trip for (%v.ToWire())) != %v", x, v) {
```

File: gen/type.go [hunk 48]
```diff
@@ -91,6 +91,15 @@ func isReferenceType(spec compile.TypeSpec) bool {
 	}
 }
 
+// Used to allow special treatment of slices in ToWire and treat nil as a valid slice.
+func isListType(spec compile.TypeSpec) bool {
+	// Lookup aliases for our type.
+	spec = compile.RootTypeSpec(spec)
+	_, isList := spec.(*compile.ListSpec)
+
```
Comment (abhinav) [inline gen/type.go]: nit: no empty line here

```suggestion
```
